### PR TITLE
Optimize SDK resolution in dotnet build

### LINKED ISSUE #9506: Project SDKs resolution order
I'm doing some tests with the project sdks and I want to override the sdk used specifying the version inside the project file.
Reading the documentation https://learn.microsoft.com/en-us/visualstudio/msbuild/how-to-use-project-sdk?view=vs-2022#how-project-sdks-are-resolved I expected that the override should work, but when I try it in VS I get this error
```
 warning MSB4241: The SDK reference "MSTest.Sdk" version "3.2.1-dev" was resolved to version "8.0.100" instead.  You could be using a different version than expected if you do not update the referenced version to match.
 warning MSB4240: Multiple versions of the same SDK "MSTest.Sdk" cannot be specified. The previously resolved SDK version "8.0.100" from location "...\Contoso.Tests.csproj" will be used and the version "3.2.1-dev" will be ignored.
 warning MSB4240: Multiple versions of the same SDK "MSTest.Sdk" cannot be specified. The previously resolved SDK version "8.0.100" from location "...\Contoso.Tests.csproj" will be used and the version "3.2.1-dev" will be ignored.
```
custom SDK is installed in C:\Program Files\dotnet\sdk\8.0.100\Sdks\MSTest.Sdk and I've specified `<Project Sdk="MSTest.Sdk/3.2.1-dev">` in the project

Is it expected?

I don't see the same issue if I build using `dotnet build` in the command line and I see the package(`3.2.1-dev`) correctly restored with nuget.

cc: @rainersigwald @ladipro  @Evangelink

### PR DESCRIPTION
**Pull request by ladipro**
Fixes #9506

### Context

In `dotnet build`, resolving in-box Sdk's has suboptimal performance. Resolver assemblies are loaded into the process even in cases where the project asks for Sdk's like `Microsoft.NET.Sdk`, which can be looked up trivially. And in fact, they are looked up trivially but only after resolver assemblies have been loaded and invoked.

The inefficiency is shown by the recently added logging:

image

There is also a subtle and confusing functional difference between `MSBuild.exe` and `dotnet build` where a NuGet package may take precedence over an in-box Sdk in `dotnet build`, but the same is not possible in `MSBuild.exe`.

### Changes Made

Made the default resolver handle the Sdk resolution first, before external resolvers are consulted. This is better perf-wise and aligns with `MSBuild.exe`, please see the large code comment for details.

This eliminates all but the last log event: `The SDK "Microsoft.NET.Sdk" was successfully resolved by the "DefaultSdkResolver" resolver to location "C:\Program Files\dotnet\sdk\9.0.100-alpha.1.24067.2\Sdks\Microsoft.NET.Sdk\Sdk" and version "".`

### Testing

Existing and new unit tests, manual testing of various builds.

Perf testing: `dotnet build` that builds a project depending on in-box Sdks is ~10 ms faster and loads 5 fewer assemblies, as long as it's built with `MSBuildEnableWorkloadResolver=false` (see below).

### Notes

While we don't have to ask the `MSBuildWorkloadSdkResolver` and `NuGetSdkResolver` to resolve in-box Sdk's anymore, this change is unfortunately not avoiding these resolvers in most builds. They are still loaded to resolve the special `Microsoft.NET.SDK.WorkloadAutoImportPropsLocator` Sdk, unless workloads are disabled.

### CODE REVIEW

**Comment by ladipro** [comment]
> While we don't have to ask the MSBuildWorkloadSdkResolver and NuGetSdkResolver to resolve in-box Sdk's anymore, this change is unfortunately not avoiding these resolvers in most builds. They are still loaded to resolve the special Microsoft.NET.SDK.WorkloadAutoImportPropsLocator Sdk, unless workloads are disabled.

@dsplaisted, do you think we can avoid resolving `Microsoft.NET.SDK.WorkloadAutoImportPropsLocator` when evaluating projects that don't use workloads? It would save 5 assembly loads + running some code, in hot scenarios measured at ~10 ms of overhead.

File: src/Framework/Traits.cs [hunk 1]
```diff
@@ -316,11 +316,6 @@ public bool LogTaskInputs
         /// </summary>
         public readonly bool DisableSdkResolutionCache = Environment.GetEnvironmentVariable("MSBUILDDISABLESDKCACHE") == "1";
 
-        /// <summary>
-        /// Disable the NuGet-based SDK resolver.
-        /// </summary>
-        public readonly bool DisableNuGetSdkResolver = Environment.GetEnvironmentVariable("MSBUILDDISABLENUGETSDKRESOLVER") == "1";
-
```
Comment (ladipro) [inline src/Framework/Traits.cs:323]: Removing an unused field here.

**Comment by dsplaisted** [comment]
> > While we don't have to ask the MSBuildWorkloadSdkResolver and NuGetSdkResolver to resolve in-box Sdk's anymore, this change is unfortunately not avoiding these resolvers in most builds. They are still loaded to resolve the special Microsoft.NET.SDK.WorkloadAutoImportPropsLocator Sdk, unless workloads are disabled.
> 
> @dsplaisted, do you think we can avoid resolving `Microsoft.NET.SDK.WorkloadAutoImportPropsLocator` when evaluating projects that don't use workloads? It would save 5 assembly loads + running some code, in hot scenarios measured at ~10 ms of overhead.

There's not an easy way to do this, because that SDK is imported before we know whether workloads are going to be used.  Even if we got rid of that import, we would still then import `Microsoft.NET.SDK.WorkloadManifestTargetsLocator`, which we need to import to determine if a workload is used, and probably would have the same perf characteristics as the PropsLocator (both of them need to figure out what workload manifests to load, so right now the PropsLocator is the one that takes that hit, but TargetsLocator would have to do basically the same thing if PropsLocator hadn't done it).

I think to speed this up we would need to speed up the resolver itself, or find a way to cache its data better.

File: src/Framework/Traits.cs (hunk 1 above)
Comment (Forgind) [reply src/Framework/Traits.cs:323]: Should we just remove it, or should we make it do something? I imagine someone wanted it at some point, so I'm curious why making it not do anything is ok.

File: src/Build/BackEnd/Components/SdkResolution/SdkResolverService.cs [hunk 2]
```diff
@@ -118,6 +118,53 @@ public virtual void ClearCaches()
...
+            // 1. DefaultSdkResolver checks the Sdks subdirectory of our SDK installation. Note that the work of resolving the
+            //    SDK version using machine-wide state and global.json (step 1.1. in `MSBuild.exe` above) has already been done
+            //    by the `dotnet` muxer. We know which SDK (capital letters) we are in, so the in-box Sdk lookup is trivial.
+            // 2. If no match, Microsoft.NET.Sdk.WorkloadMSBuildSdkResolver is loaded and asked to resolve the Sdk required by the project.
+            //    2.1. It checks installed workloads.
+            // 3. If no match still, Microsoft.Build.NuGetSdkResolver is loaded and asked to resolve the Sdk.
+            //
+            // Overall, while Sdk resolvers look like a general plug-in system, there are good reasons why some of the logic is hard-coded.
+            // It's not really meant to be modified outside of very special/internal scenarios.
+#if NETCOREAPP
```
Comment (Forgind) [inline src/Build/BackEnd/Components/SdkResolution/SdkResolverService.cs:147]: I may not understand how all this works properly, but I think we load SdkResolvers from manifests, right? Could we tweak the manifest instead of changing MSBuild here?

Comment (ladipro) [reply src/Build/BackEnd/Components/SdkResolution/SdkResolverService.cs:147]: Tweaking the manifests is something we should look into as well. One important thing currently missing is the resolver priority. To know the order, in which resolvers should be called, we have to load and instantiate all of them and read their `Priority` property. I think it would be reasonable to require the priority to be specified declaratively in the manifest - which would let us skip loading the NuGet resolver when the Sdk is resolved by the MSBuildSdk/Workloads resolver, for the common example. With this in place, the Default resolver could have its priority conditionally set depending on if we're compiling Core or Framework MSBuild.

**Comment by ladipro** [comment]
> > @dsplaisted, do you think we can avoid resolving `Microsoft.NET.SDK.WorkloadAutoImportPropsLocator` when evaluating projects that don't use workloads? It would save 5 assembly loads + running some code, in hot scenarios measured at ~10 ms of overhead.
> 
> There's not an easy way to do this, because that SDK is imported before we know whether workloads are going to be used. Even if we got rid of that import, we would still then import `Microsoft.NET.SDK.WorkloadManifestTargetsLocator`, which we need to import to determine if a workload is used, and probably would have the same perf characteristics as the PropsLocator (both of them need to figure out what workload manifests to load, so right now the PropsLocator is the one that takes that hit, but TargetsLocator would have to do basically the same thing if PropsLocator hadn't done it).
> 
> I think to speed this up we would need to speed up the resolver itself, or find a way to cache its data better.

Thank you. Can we realistically defer these imports until `WorkloadMSBuildSdkResolver` is invoked to resolve a "real" Sdk? I may not be understanding the full complexity but it looks as though if all (user-specified) Sdk's are resolved by simply finding the matching directory under `Sdks`, then we know for sure no workload is used. Only if we end up actually calling `WorkloadMSBuildSdkResolver`, there is a possibility it is a workload (or it can be a NuGet-delivered Sdk - we have to bite the bullet either way).

File: src/Framework/Traits.cs (hunk 1 above)
Comment (ladipro) [reply src/Framework/Traits.cs:323]: This is likely a leftover from when the NuGet resolver was part of this repo. The environment variable is read in the resolver and presumably doing what it was doing here:
https://github.com/NuGet/NuGet.Client/blob/566e2853b8537e8b484844fad32b3f6073ba3167/src/NuGet.Core/Microsoft.Build.NuGetSdkResolver/NuGetSdkResolver.cs#L34

**Comment by dsplaisted** [comment]
> Thank you. Can we realistically defer these imports until `WorkloadMSBuildSdkResolver` is invoked to resolve a "real" Sdk? I may not be understanding the full complexity but it looks as though if all (user-specified) Sdk's are resolved by simply finding the matching directory under `Sdks`, then we know for sure no workload is used. Only if we end up actually calling `WorkloadMSBuildSdkResolver`, there is a possibility it is a workload (or it can be a NuGet-delivered Sdk - we have to bite the bullet either way).

@ladipro A project doesn't explicitly specify an MSBuild SDK import to depend on a workload.  Workloads can be enabled by any MSBuild property, and that mapping is defined in the WorkloadManifest.targets file of each workload manifest.  For example, if the target platform is Android, then the Android workload will be required, and the `RunAOTCompilation` property enables the wasm-tools workload for Blazor projects.  How this works is covered in this design doc: https://github.com/dotnet/designs/blob/main/accepted/2020/workloads/workload-resolvers.md.

Basically we can't know whether a project uses a workload or not without first running the workloads resolver and processing the .targets files that come from the workload manifests.

**Comment by ladipro** [comment]
> @ladipro A project doesn't explicitly specify an MSBuild SDK import to depend on a workload. Workloads can be enabled by any MSBuild property, and that mapping is defined in the WorkloadManifest.targets file of each workload manifest. For example, if the target platform is Android, then the Android workload will be required, and the `RunAOTCompilation` property enables the wasm-tools workload for Blazor projects. How this works is covered in this design doc: https://github.com/dotnet/designs/blob/main/accepted/2020/workloads/workload-resolvers.md.
> 
> Basically we can't know whether a project uses a workload or not without first running the workloads resolver and processing the .targets files that come from the workload manifests.

Ah, I was completely off base then. Ok, so the next best thing we can do is have resolvers statically declare their priority so at least the NuGet resolver is pay-for-play - loaded and called only when actually needed. That should be mostly MSBuild work. Anyway, thank you!

**Review by JanKrivanek — APPROVED** [review]

**Review by AR-May — APPROVED** [review]
